### ISSUE TITLE
S24LE distortion issue

### ISSUE THREAD

**Issue by alex-vyverman** [HEnquist/camilladsp#102]
First and foremost my sincere congratulations for this DSP project. It is absolutely remarkable!

I experience an issue though. To test Camilla properly, I made a simple stereo in to stereo out pipeline without any processing
`

> devices:
>   samplerate: 96000
>   chunksize: 1024
>   capture:
>     type: Alsa
>     channels: 2
>     device: "hw:0,0"
>     format: S16LE
>   playback:
>     type: Alsa
>     channels: 2
>     device: "hw:0,0"
>     format: S16LE
> 
> mixers:
>   stereo:
>     channels:
>       in: 2
>       out: 2
>     mapping:
>       - dest: 0
>         sources:
>           - channel: 0
>             gain: -6
>             inverted: false
>       - dest: 1
>         sources:
>           - channel: 1
>             gain: -6
>             inverted: false
> 
> pipeline:
>   - type: Mixer
>     name: stereo

`

This works fine without any issues.

But is I change S16_LE to S24_LE, like this:

> `---
> devices:
>   samplerate: 96000
>   chunksize: 1024
>   capture:
>     type: Alsa
>     channels: 2
>     device: "hw:0,0"
>     format: S24LE
>   playback:
>     type: Alsa
>     channels: 2
>     device: "hw:0,0"
>     format: S24LE
> 
> mixers:
>   stereo:
>     channels:
>       in: 2
>       out: 2
>     mapping:
>       - dest: 0
>         sources:
>           - channel: 0
>             gain: -6
>             inverted: false
>       - dest: 1
>         sources:
>           - channel: 1
>             gain: -6
>             inverted: false
> 
> pipeline:
>   - type: Mixer
>     name: stereo
> `

I get very strong distortion and clipping occurring as if the process is trying to read 24bit info as only the lower 16bits, or some similar kind of bit shift causing massive clipping. 

I've tried to debug the capabilities of my interface by using:

`arecord -c 2 -f S24_LE -r 96000 | aplay -c 2 -f S24_LE -r 96000`

Basically piping arecord into aplay at 96/24. This worked fine without any distortion. By which I assume my interface is S24_LE capable on both ends.

My setup is:

Raspberry Pi 4, raspberry Pi OS 
uname -a:
`Linux audioRpi4 5.10.17-v7l+ #1403 SMP Mon Feb 22 11:33:35 GMT 2021 armv7l GNU/Linux`

camilladsp -V
`CamillaDSP 0.5.0-beta6`

Audio interface:
HifiBerry Digi+ I/O using ALSA over I2S

Source signal is coming over SPDIF from an RME Fireface 400 interface set to 96KHz samplerate.

Anyway of figuring out where the distortion is coming from?

**Comment by HEnquist** [comment]
Thanks for reporting this! I'll take a look.
Could I ask you to try running with 24 bits on only one device at a time? That would tell if the problem comes from the capture or playback side (or both).

**Comment by HEnquist** [comment]
One more thing, I don't have any device that supports S24LE. 
Could you record a short (small fraction of a second) bit of audio data with that arecord command and attach it here? Something like `arecord -c 2 -f S24_LE -r 96000 > dump.raw`

**Comment by HEnquist** [comment]
Ok I found the problem. It doesn't handle negative values properly for S24LE. This has most likely never been working in any version, seems like this format isn't used much.. Expect a fixed version soon!

**Comment by alex-vyverman** [comment]
That's great! Thanks for your swift and dedicated response.

I have been experimenting a bit. I couldn't do different bit rates between input and output and get it to work.

Weirdly enough I could only do S16_LE and S24_LE with this card at 96k
My goal was to do at least 24/96 with my setup, but it wouldn't run 32/96 at first.
...and now it miraculously does all of a sudden.

It seems this suddenly started working after switching from "hw" to "plughw", and then back to "hw". I don't know ALSA enough to understand if this makes sense in any way. But 32bit int is an even better option for my case, so no emergency on my part as long as it decides to keep working at these rates.

But it would make sense to fix it for those who want to use S24_LE, at least this thread makes sure people are aware of an issue.

Thanks again for this amazing piece of software!

**Comment by HEnquist** [comment]
I think I have fixed it, could you give this new version a try?
https://github.com/HEnquist/camilladsp/releases/tag/v0.5.0-s24test

**Comment by alex-vyverman** [comment]
I just found the release. Works very well at S24LE now!

**Comment by HEnquist** [comment]
Excellent! Thanks for helping out :)
